### PR TITLE
HBHE: mahi speed up

### PR DESCRIPTION
**Pull request by mariadalfonso**
Code cleanup, followed from V.Innocente's suggestions, resulted in a 10% speed improvement. 
Tested on data 2018-RunB, see igproof below (at commit dfc08dbd4292947382a16b7df9df550c864a8ea7 ).

No changes are expected in the HCAL quantities, neither energy/time.

- - - - - - - - - - - OLD - - - - - - - - - - - - - - - - - - - -
```
            1.5  .........       1.77 / 2.13         SimpleHBHEPhase1Algo::reconstruct(HBHEChannelInfo const&, HcalRecoParam const*, HcalCalibrations const&, bool) [67]
[76]        1.5       1.77       0.07 / 1.69       MahiFit::phase1Apply(HBHEChannelInfo const&, float&, float&, bool&, float&) const
            1.3  .........       1.62 / 1.62         MahiFit::doFit(std::array<float, 3ul>&, int) const [85]
            0.1  .........       0.07 / 0.07         MahiFit::resetWorkspace() const [786]
```

```
            1.3  .........       1.62 / 1.77         MahiFit::phase1Apply(HBHEChannelInfo const&, float&, float&, bool&, float&) const [76]
[85]        1.3       1.62       0.03 / 1.59       MahiFit::doFit(std::array<float, 3ul>&, int) const
            1.0  .........       1.23 / 1.23         MahiFit::minimize() const [120]
            0.2  .........       0.20 / 0.20         MahiFit::updatePulseShape(double, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<double, -1, -1, 0, 19, 19>&) const [426]
            0.1  .........       0.16 / 0.16         MahiFit::calculateArrivalTime() const [489]
```

- - - - - - - - - - -   NEW - - - - - - - - - - - - - - - - - - - -
```
            1.3  .........       1.56 / 1.95         SimpleHBHEPhase1Algo::reconstruct(HBHEChannelInfo const&, HcalRecoParam const*, HcalCalibrations const&, bool) [70]
[83]        1.3       1.56       0.05 / 1.50       MahiFit::phase1Apply(HBHEChannelInfo const&, float&, float&, bool&, float&) const
            1.2  .........       1.48 / 1.48         MahiFit::doFit(std::array<float, 3ul>&, int) const [86]
            0.0  .........       0.02 / 0.02         MahiFit::resetWorkspace() const [1301]

```

```
            1.2  .........       1.48 / 1.56         MahiFit::phase1Apply(HBHEChannelInfo const&, float&, float&, bool&, float&) const [83]
[86]        1.2       1.48       0.07 / 1.41       MahiFit::doFit(std::array<float, 3ul>&, int) const
            1.0  .........       1.19 / 1.19         MahiFit::minimize() const [121]
            0.1  .........       0.14 / 0.14         MahiFit::updatePulseShape(double, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<double, -1, -1, 0, 19, 19>&\
) const [507]
            0.1  .........       0.07 / 0.07         MahiFit::calculateArrivalTime() const [754]

```

### CODE REVIEW

**Comment by slava77** [comment]
>  No changes are expected in the HCAL quantities.

is incompatible with 

> Reco comparison results: 7919 differences found in the comparisons

At a glance (in the PR), the differences are likely due to changes in double and float precision.
To make sure that there are no actually significant changes, please provide hit-by-hit comparisons for MAHI and M2 for some events with pileup as well as without pileup (to check both the real pulse and noise dominated environments). It would be good to have coverage up to a TeV regime.

**Comment by mariadalfonso** [comment]
@slava77

the difference thisPR/release comes from the mahi arrival time estimation 
http://dalfonso.web.cern.ch/dalfonso/HCAL/PR25982/recHitTime_PR25982.png
but actually seems that this PR goes in the direction of bug-fixing it :)
I suspect comes from the initialization of some of the nnlsWork_ matrices, I will understand what exactly is the right line.

@hum

**Comment by slava77** [comment]
>  I suspect comes from the initialization of some of the nnlsWork_ matrices, I will understand what exactly is the right line.

Please update the PR description after that.
Presently it reads as if the PR should introduce no changes in the physics output.

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 1]
```diff
@@ -155,26 +155,32 @@ void MahiFit::doFit(std::array<float,3> &correctedOutput, int nbx) const {
   if (dynamicPed_) nnlsWork_.bxs[nnlsWork_.nPulseTot-1] = pedestalBX_;
 
   nnlsWork_.pulseMat.setZero(nnlsWork_.tsSize,nnlsWork_.nPulseTot);  
+  nnlsWork_.pulseDerivMat.setZero(nnlsWork_.tsSize,nnlsWork_.nPulseTot);
+
+  std::array<FullSampleVector, MaxPVSize> pulseShapeArray;
```
Comment (VinInn) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: there is no need for the array anymore actually....
just remove [iBX] as well for these two...

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 2]
```diff
@@ -284,7 +284,7 @@ void MahiFit::updatePulseShape(double itQ, FullSampleVector &pulseShape, FullSam
   //with previous SOI=TS4 case assumed by psfPtr_->getPulseShape()
   int delta =  4 - nnlsWork_. tsOffset;
 
-  auto invDt = ( 1./2*nnlsWork_.dt );
+  auto invDt = 1./ ( 2*nnlsWork_.dt );
```
Comment (VinInn) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: to over-do-it you can remove the 0.5 below and here have 0.25/nnlsWork_.dt ...

**Comment by mariadalfonso** [comment]
@slava77 @perrotta

fix the problem with the arrivalTime in this PR.
we can keep this PR as pure technical and we should not observe any change in HCAL quantities.
we can restart the test.

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by mariadalfonso** [comment]
"Reco comparison results: 7762 differences found in the comparisons"

differences are pretty minimal, probably due to roundings float vs double.

Most cases have: no changes in the HCAL-recHits + some minimal in the downstreams

**Comment by slava77** [comment]
I'd like to repeat the earlier comment:

>  To make sure that there are no actually significant changes, please provide hit-by-hit comparisons for MAHI and M2 for some events with pileup as well as without pileup (to check both the real pulse and noise dominated environments). It would be good to have coverage up to a TeV regime.

**Comment by mariadalfonso** [comment]
@slava77

Here some scatter plot for both M2 and Mahi,
http://dalfonso.web.cern.ch/dalfonso/HCAL/PR25982.pdf
Conclusions as before, the difference are limited to rounding effects

**Comment by slava77** [comment]
> @slava77
> 
> Here some scatter plot for both M2 and Mahi,
> http://dalfonso.web.cern.ch/dalfonso/HCAL/PR25982.pdf
> Conclusions as before, the difference are limited to rounding effects

50 GeV range is a bit too small.
 The low energy range is almost invisible: may I ask for a log10 on x/y axis.

The timing plot has some more significant deviations. Are these clearly from numerical changes?
screen shot 2019-02-28 at 9 01 22 am

**Comment by mariadalfonso** [comment]
@slava77

I updated the validation plots here (i.e. log xy)
http://dalfonso.web.cern.ch/dalfonso/HCAL/PR25982.pdf

I rolled back some of the float vs double changes to restore the ancient M2.

"My conclusions" as before: 
1. MAHI energies are good, 
2. MAHI arrival time almost invisible numerical changes less than before (note this variable is completely random at the moment and only used to check the compatibility with +- 25ns in the beam halo filter); 
3. M2 energy/time some difference but negligible (but this algo is completely dismissed and not in any sequence)

Can cmsbot re-run the test and hopefully we can merge (note jenkin output will be the similar). I would like to have this in for the legacy 10_6 and for the baseline HLT Run3 studies.

To put in prospective, for the Run3, HCAL has 1.7 more channels, so my goal is to gain for HLT a factor 1.7, this PR make up for a free 10% , the road is still long and much more than rounding effects will have to be provided and accepted.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25982	4b3a7c4
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show small differences starting from hbhe rechits and propagating downstream; the differences are consistent with numerical precision level  changes
- local tests with flat-PT dijet 15 to 7TeV with PU50 shows expected behavior, in line with the plots provided in https://github.com/cms-sw/cmssw/pull/25982#issuecomment-469417446 
    - some example numerology for HE: out of 6,880,364 rechits, there are 1,116,150 with differences in energy. Only 155 have |delta|/sum > 0.0001, of these 0 have |delta| > 0.01 MeV. Most differences are in hits with very small energy values (fits of noise), where fractional changes within numerical precision are more likely.
    - time in hbheprereco is down by about 7%

**Comment by VinInn** [comment]
so it was eventually decided that moving float to double (in a fully double algorithm) is bad for physics, wasn't it?

**Comment by mariadalfonso** [comment]
@VinInn I will try to move in the future as much as possible those double to float.

**Comment by mariadalfonso** [comment]
@tocheng can you sign off , the changes are related to the double --> float in the TimeSlew class that is in the CalibCalorimetry package

**Comment by slava77** [comment]
On 3/6/19 11:55 PM, Vincenzo Innocente wrote:
> so it was eventually decided that moving float to double (in a fully
> double algorithm) is bad for physics, wasn't it?
> 

I didn't keep track of all the double->float and float->double changes
made while this PR was developing in response to my questions/requests
for investigation of origins of the differences.
Which edits are you referring to that did not do a float -> double revert?

Are there any opportunities for vectorization?
If so, then a simple switch to double will hurt the timing, especially
if the floats were acceptable for physics already.

**Comment by VinInn** [comment]
no vectorization
algo are fully coded in double, some parameters are in float, so there are conversions float->double that can be avoided.

Anyhow, as @mariadalfonso commented, the best is to migrate everything to float and keep only the critical part of the fit in double...

**Comment by slava77** [comment]
On 3/7/19 5:29 AM, Vincenzo Innocente wrote:
> no vectorization
> algo are fully coded in double, some parameters are in float, so there
> are conversions float->double that can be avoided.
> 
> Anyhow, as @mariadalfonso
> commented, the best is to migrate everything to float and keep only the
> critical part of the fit in double...
> 
>

OK.
A while back I checked the stability of M2 and it seemed important to
compute the chi2 in double, while all other code could be in float.
I seem to recall that there was enough math unrelated to chi2 that using
float there was justified even despite the extra cost of float -> double
conversion.
I wonder if MAHI is more stable adn could indeed work with just float.

**Comment by slava77** [comment]
@pohsun @tocheng 
ALCA signature is missing for this PR.
Please check and comment or sign.
Thank you.

**Comment by pohsun** [comment]
@mariadalfonso 
Are you planning to implement those 'float' in the non-critical parts soon?
Or I could just sign-off?

**Comment by mariadalfonso** [comment]
@pohsun 
I consider this PR completed.

**Comment by pohsun** [comment]
@mariadalfonso Thanks for clarification.
+1

**Comment by slava77** [comment]
> @mariadalfonso Thanks for clarification.
> +1

it seems like a "+1" is accepted only if it's in the beginning of the message.
@pohsun please edit your original message or post another "+1"
Thank you.

**Comment by fabiocos** [comment]
@slava77 @VinInn so do I understand correctly that you suggest/agree to test the version of MAHI proposed by this PR, and possibly consider further updates later on?

**Comment by VinInn** [comment]
@fabiocos , yes

**Comment by fabiocos** [comment]
+1

**Pull request merged**